lib/metadata/Metadata.hpp

```diff
@@ -0,0 +1,186 @@
+/**
+ * @file
+ * @brief Contains Metadata class.
+ */
+
+#ifndef METADATA_HPP
+#define METADATA_HPP
+
+#include <algorithm>
+#include <cassert>
+#include <cstdint>
+#include <initializer_list>
+#include <iostream>
+#include <string>
+#include <type_traits>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace chord {
+
+// The Metadata class is roughly similar to a Python dictionary. It
+// holds typed key/value pairs. The keys are strings, and the values
+// are either boolen, integer, float, string, or a vector of these.
+//
+// Values cannot be overridden once set. This is a write-once data
+// structure.
```

Comment:
I'm not sure about making this write-once and not copyable.  There are places in the pipeline where it's nice to add or update a few parameters, but otherwise leave the rest of the metadata data intact while passing it on to the next stage.   There are also place the metadata gets copied between two branches of the pipeline and treated differently in each branch.

Reply:
Adding new attributes is possible, changing attribute values is not. We can add this, but we don't do that anywhere at the moment, and so it sounds like a nice design principle avoiding accidental errors.

Copying makes sense. Not sure whether I want a copy constructor for this or a more explicit operations since it's expensive. It's probably fine to stick with C++ semantics here and just allow a copy constructor.